### PR TITLE
 Adding dimensions_to_exclude to keep their positions

### PR DESCRIPTION
**Pull request by ivankulman**
Resolves https://github.com/cubewise-code/optimus-py/issues/27

1. Adding dimensions to exclude so they keep their position in the stored dim order
2. Added a check to prevent dimensions with strings from being put to the last position (because it's irreversible may break future swaps with this dim)
3. executionMode is a separate file
4. minor renames to reflect the intention behind the variable

### CODE REVIEW

**Comment by ivankulman** [comment]
I would also strongly recommend running a syntax check like `black` on top of the code base to make it consistent for everyone involved.

File: executors.py [hunk 1]
```diff
@@ -22,20 +22,6 @@ def swap_random(order: list) -> List[str]:
     return swap(order, i1, i2)
 
 
-class ExecutionMode(Enum):
```
Comment (ivankulman) [inline executors.py:25]: split to separate file

File: executors.py [hunk 2]
```diff
@@ -178,34 +195,42 @@ def execute(self) -> List[PermutationResult]:
             dimensions.remove(self.dimensions[-1])
 
         # iteration through positions like: n, 0, n-1, 1, n-2, 2, ...
-        for iteration, position in enumerate(chain(*zip(reversed(range(len(dimensions))), range(len(dimensions))))):
+        for iteration, target_position in enumerate(chain(*zip(reversed(range(len(dimensions))), range(len(dimensions))))):
```
Comment (ivankulman) [inline executors.py:198]: dimension swaps from original_position to target_position, makes it easier to understand what the swaps are doing

File: executors.py [hunk 3]
```diff
@@ -178,34 +195,42 @@ def execute(self) -> List[PermutationResult]:
...
-                best_order = sorted(
-                    results_per_dimension,
-                    key=lambda r: r.median_query_time(self.view_name))[0]
-
-            resulting_order = list(best_order.dimension_order)
-            dimension_pool.remove(resulting_order[position])
+                dimension_target = resulting_order[target_position]
+
+                if (not self._check_swap_dim_with_str_to_last_position(dimension, target_position)
+                    and dimension_target in dimension_pool):
```
Comment (ivankulman) [inline executors.py:213]: while new line 208 checks for dimension in the allowed dims, this check is for the target dims, so we don't swap dims if we don't want to

File: optimuspy.py [hunk 4]
```diff
@@ -164,49 +164,46 @@ def main(instance_name: str, cube_name: str, view_name: str, process_name: str,
             write_vmm_vmt(tm1, cube_name, "1000000", "1000000")
 
             logging.info(f"Starting analysis for cube '{cube_name}'")
-            original_dimension_order = tm1.cubes.get_storage_dimension_order(cube_name=cube_name)
-            logging.info(f"Original dimension order for cube '{cube_name}' is: '{original_dimension_order}'")
+            storage_dimension_order = tm1.cubes.get_storage_dimension_order(cube_name=cube_name)
```
Comment (ivankulman) [inline optimuspy.py:167]: original to me means the "display" order, while this is the storage order

**Comment by ivankulman** [comment]
@MariusWirtz
Hi Marius, please review when you get a chance.

File: executionMode.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+from enum import Enum
```
Comment (MariusWirtz) [inline executionMode.py:1]: please rename to `execution_mode.py`

File: executors.py (hunk 3 above)
Comment (MariusWirtz) [reply executors.py:213]: Understood

File: optimuspy.py (hunk 4 above)
Comment (MariusWirtz) [reply optimuspy.py:167]: We use the term `OriginalOrder` also when presenting the results.
https://github.com/cubewise-code/optimus-py/issues/49

I think `original order` is a better term than `storage dimension order` especially when presenting the results.
Original would refer to the point in time before the optimization. I don't like the term storage dimension order for the initial order as the storage dimension order mutates over time.

I wonder if we can agree on an alternative term to use for the variable and in the output. 
How about `InitialDimensionOrder`?

File: optimuspy.py [hunk 6]
```diff
@@ -164,49 +164,46 @@ def main(instance_name: str, cube_name: str, view_name: str, process_name: str,
...
                     else:
                         logging.info(f"Best order for cube '{cube_name}': {best_order}")
-                        tm1.cubes.update_storage_dimension_order(cube_name, original_dimension_order)
+                        tm1.cubes.update_storage_dimension_order(cube_name, storage_dimension_order)
                         logging.info(
-                            f"Restored original dimension order for cube '{cube_name}' to {original_dimension_order}")
-
-            except:
-                logging.error("Fatal error", exc_info=True)
+                            f"Restored original dimension order for cube '{cube_name}' to {storage_dimension_order}")
```
Comment (MariusWirtz) [inline optimuspy.py:203]: Ideally we drop the term "original dimension order" consistently in the project and substitute it with something everyone can agree with :)

**Comment by MariusWirtz** [comment]
Nice work @ivankulman 

I agree with substituting the term Original Order. It is confusing.
Please add the name changes and then we can merge.

File: optimuspy.py (hunk 4 above)
Comment (MariusWirtz) [reply optimuspy.py:167]: I think we should go with `InitialDimensionOrder`

**Review by ivankulman — COMMENTED** [review]
added changes as per the feedback
1. original_dimension_order -> initial_dimension_order
2. executionMode to execution_mode

**Pull request merged**